### PR TITLE
elpy-shell: bug fix and improvements

### PR DESCRIPTION
**Pull request by rgemulla**
Fixes bug #1225.

Also adds functionality to evaluate code cells. This can be used to send custom sets of lines to the shell (by default, the set of lines is delimited by ``###`` lines). Full documentation is in PR #1227.

### CODE REVIEW

File: elpy-shell.el [hunk 1]
```diff
@@ -884,71 +915,118 @@ below point and send the group around this statement."
...
+        (end (save-excursion
+               (forward-line)
+               (if (re-search-forward elpy-shell-cell-boundary-regexp nil t)
+                   (forward-line -1)
+                 (end-of-buffer))
+               (end-of-line)
+               (point))))
+    (if beg
+        (progn
+          (elpy-shell--flash-and-message-region beg end)
```
Comment (galaunay) [inline elpy-shell.el]: Is it not possible to use `elpy-shell-send-region-or-buffer` here to avoid redundancy ?

**Comment by galaunay** [comment]
Is this a wanted behaviour to send the code until the end of the buffer in there is not ending block ?

Minor issue maybe, the code block is not sent if the cursor is on the code block ending line.
i.e. on the third line here:
```Python
###
print("Hello world")
###
```

File: elpy-shell.el (hunk 1 above)
Comment (rgemulla) [reply elpy-shell.el]: Possible, but then we'd have to store, change, and restore the region. I'd vote to keep it as is. The entire code-sending parts should probabily be simplified into:

1. Navigation commands for start/end
2. Code sending commands (automatically build from navigation commands)

I'll do this at some point. The main reasoning is that users can then also use the navigation commands directly (e.g., to jump to the next codecell etc.)

**Comment by rgemulla** [comment]
Yes. The semantics is: The current cell starts at the cell boundary on or above the line of point and ends at the cell boundary below point (or end of buffer, if none). A cell is executed only if the starting boundary matches a codecell beginning. In your example above, the second ```###``` acts as a cell boundary and as a beginning of a new codecell.

This makes it easy, for example, to support the cells of a notebook exported from Jupyter directly (where the beginning is "# In[.*]:" or "# <codecell>"). It also avoids accidential execution of non-codecells (where the begining is "# Out[]:" or "# <markdowncell>" etc.). This is the default behaviour.

What is boundary and what codecell beginning can be customized via the regexps.

**Comment by galaunay** [comment]
I understand now, thank you for the clarifications.

**Comment by rgemulla** [comment]
To me, the PR is ready. @galaunay Agreed?

File: elpy-shell.el (hunk 1 above)
Comment (galaunay) [reply elpy-shell.el]: This indeed seems a good idea is the number of functions sending code parts increases.

**Comment by galaunay** [comment]
The mix between bug fix and code cell sending make it a little hard to read, but this seems ok for me.

**Comment by rgemulla** [comment]
Yes, I'll try not to do that again. Thanks!

**Pull request merged**